pkg/scheduler/plugins/reservation/transformer.go

```diff
@@ -22,210 +22,361 @@ import (
-		"pod", klog.KObj(pod))
+	cycleState.Write(stateKey, state)
 
-	// initialize a state for matched reservations.
-	// if the pod match any reservations, it stores the matched reservations meta for the pod, mapping from nodeName
-	// to reservationInfo
-	cycleState.Write(preFilterStateKey, state)
+	klog.V(4).Infof("Pod %v has %d matched reservations and %d unmatched reservations before PreFilter", klog.KObj(pod), len(state.matched), len(state.unmatched))
+	return nil, false
+}
 
-	if state.skip {
-		klog.V(5).InfoS("BeforePreFilter skips for no reservation matched", "pod", klog.KObj(pod))
-		return nil, false
+func (pl *Plugin) AfterPreFilter(_ frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod) error {
+	state := getStateData(cycleState)
+	for nodeName, reservationInfos := range state.unmatched {
+		nodeInfo, err := pl.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+		if err != nil {
+			continue
+		}
+		if nodeInfo.Node() == nil {
+			continue
+		}
+
+		if err := pl.restoreUnmatchedReservations(cycleState, pod, nodeInfo, reservationInfos, true); err != nil {
+			return err
+		}
 	}
 
-	// skip pod pre-filter of affinities/anti-affinities, topology constrains
-	return preparePreFilterPod(pod), true
-}
+	if !reservationutil.IsReservePod(pod) {
+		for nodeName, reservationInfos := range state.matched {
+			nodeInfo, err := pl.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+			if err != nil {
+				continue
+			}
+			if nodeInfo.Node() == nil {
+				continue
+			}
+
+			// NOTE: After PreFilter, all reserved resources need to be returned to the corresponding NodeInfo,
+			// and each plugin is triggered to adjust its own StateData through RemovePod, RemoveReservation and AddPodInReservation,
+			// and adjust the state data related to Reservation and Pod
+			if err := pl.restoreMatchedReservations(cycleState, pod, nodeInfo, reservationInfos, true); err != nil {
+				return err
+			}
+		}
+	}
 
-func (p *Plugin) AfterPreFilter(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod) error {
 	return nil
 }
 
-func (p *Plugin) BeforeFilter(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) (*corev1.Pod, *framework.NodeInfo, bool) {
-	// do not transformer if not reserve state (where we should check if pod match any reservation on node)
-	if reservationutil.IsReservePod(pod) {
-		return nil, nil, false
-	}
-	// abort if node is not found
+func (pl *Plugin) BeforeFilter(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) (*corev1.Pod, *framework.NodeInfo, bool) {
 	node := nodeInfo.Node()
 	if node == nil {
 		return nil, nil, false
 	}
+	// Generation equals -1 means that the Reservation return logic has been executed once, but it should not happen.
+	if nodeInfo.Generation == -1 {
```

Comment:
how does nodeInfo.Generation=-1 works

Reply:
The purpose of updating Generation is to reduce the calculation amount of nodeinfo snapshot. Most nodes may not change during the scheduling phase, that is, the generation is the original, so this node does not need to be updated. The Generation is -1, which means that nNodeInfo has been modified and must be updated.